flight/Libraries/math/cic.c

```diff
@@ -0,0 +1,160 @@
+ */
+void cic_apply_higher_order_int_stage(struct cic_filter_data *filter_data) {
+	if (cic_filter_order  >= 2) {
+		filter_data->integrateState1 += filter_data->integrateState0;
+	}
+
+	if (cic_filter_order >= 3) {
+		filter_data->integrateState2 += filter_data->integrateState1;
+	}
+}
+
+/**
+ * @brief Applying the comb stage of the CIC filter
+ * @param[in] filter_data Pointer to the CIC filter structure which stores the data / information
+ */
+void cic_apply_comb_stage(struct cic_filter_data *filter_data) {
+	if (cic_filter_order == 1) {
+		filter_data->combState0 = filter_data->integrateState0 - filter_data->integrateState_old;
+
+		if (cic_filter_ddelay == 1) {
+			filter_data->integrateState_old = filter_data->integrateState0;
+		}
+		else if (cic_filter_ddelay == 2) {
+			filter_data->integrateState_old      = filter_data->integrateState_old_temp;
+			filter_data->integrateState_old_temp = filter_data->integrateState0;
+		}
+	}
+
+	else if (cic_filter_order == 2) {
+		filter_data->combState0 = filter_data->integrateState1 - filter_data->integrateState_old;
+		filter_data->combState1 = filter_data->combState0  - filter_data->combState0_old;
+
+		if (cic_filter_ddelay == 1) {
+			filter_data->integrateState_old = filter_data->integrateState1;
+			filter_data->combState0_old = filter_data->combState0;
+		}
+		else if (cic_filter_ddelay == 2) {
+			filter_data->integrateState_old      = filter_data->integrateState_old_temp;
+			filter_data->integrateState_old_temp = filter_data->integrateState1;
+			filter_data->combState0_old      = filter_data->combState0_old_temp;
+			filter_data->combState0_old_temp = filter_data->combState0;
+		}
+	}
+
+	else if (cic_filter_order == 3) {
+		filter_data->combState0 = filter_data->integrateState2 - filter_data->integrateState_old;
+		filter_data->combState1 = filter_data->combState0  - filter_data->combState0_old;
+		filter_data->combState2 = filter_data->combState1  - filter_data->combState1_old;
+
+		if (cic_filter_ddelay == 1) {
+			filter_data->integrateState_old = filter_data->integrateState2;
+			filter_data->combState0_old = filter_data->combState0;
+			filter_data->combState1_old = filter_data->combState1;
+		}
+		else if (cic_filter_ddelay == 2) {
+			filter_data->integrateState_old      = filter_data->integrateState_old_temp;
+			filter_data->integrateState_old_temp = filter_data->integrateState2;
+			filter_data->combState0_old      = filter_data->combState0_old_temp;
+			filter_data->combState0_old_temp = filter_data->combState0;
+			filter_data->combState1_old      = filter_data->combState1_old_temp;
+			filter_data->combState1_old_temp = filter_data->combState1;
+		}
+	}
+}
+
+/**
+ * @brief Get the decimated / down-sampled output of the CIC filter
+ * @param[in] filter_data Pointer to the CIC filter structure which stores the data / information
+ * @returns Output the computed gain of the filter
+ */
+int32_t cic_get_decimation_output(struct cic_filter_data *filter_data) {
```

Comment:
You mean const pointers like
```
void cic_apply_higher_order_int_stage(const struct cic_filter_data *filter_data);
```

I have read that this might have speed improvements, so changed it for all the cic  function, and also at some
I have read that this might have speed improvements, so changed it for all the cic  function, and also at other place
```
void PIOS_MPU6000_SetGyroDownSamling(const uint8_t *gyro_downsampling)
```